File: 150to300/153_FindMinimuminRotatedSortedArray.java
/*
relfection point都是最小值，[5,1,2,3,4] ， 1 这个点就是relfection point

三个情况:
1. mid > start ， 说明从start 到 mid都是sorted，那么reflection point一定在右边
2. 不然就是上述情况， 从 mid - end都是sorted，所以reflection point在左边
3. 什么情况下是找到了reflection :
   a. nums[mid - 1] > nums[mid], 如果arraysorted，那不可能mid -1 小于 mid，出现这个可能只能是mid是最小值
   b。 nums[mid + 1] < nums[mid], 同理， mid +1只能为reflection point

    /
   /
  /
  ---------------
                /
               /
              /  
   
*/

class Solution {
    public int findMin(int[] nums) {
        if (nums == null || nums.length == 0) return -1;
        //corner case,当array只有一个数的时候
        if (nums.length == 1) return nums[0];
        int start = 0;
        int end = nums.length - 1;
        int mid = -1;
        //corner case，当array的转动pivot是num[0]，array就变成了原本的sorted array了
        if(nums[end] > nums[start]) return nums[start];
        
        while (start <= end){
            mid = (end + start) / 2;
            int midNum = nums[mid];
            int startNum = nums[start];
            int endNum = nums[end];
            /*
            1 2 3 4 5 6 7 => 4 5 6 7 1 2 3 
            */
            if (nums[mid] > nums[mid + 1]){
                return nums[mid + 1];
            }
            /*
            1 2 3 4 5 6 7  =>  6 7 1 2 3 4 5
            */
            if (nums[mid] < nums[mid - 1]){
                return nums[mid];
            }
            /*
            如果中间数大于头，说明从 start ~ mid都是sorted，那反转的地方在后面，且反转的地方才存在reflection point
            1 2 3 4 5 6 7 => 4 5 6 7 1 2 3
            */
            if (midNum > startNum){
                start = mid + 1;
            } else {//反之则是从mid  ~  end都是sorted，1 2 3 4 5 6 7  =>  6 7 1 2 3 4 5
                end = mid - 1;
            }
        }
        
        return -1;
    }
}

File: 150to300/153_FindMinimuminRotatedSortedArray.py
'''
我们要找的是pivot，也就是rotate的哪个点，那个点一定是最小值
那么这有一个特征：
1. mid + 1 小于mid: 4 5 [6] 1 2 3   所以return mid + 1
2. pivot就在mid ：  mid - 1 > mid and mid < mid + 1       5 6 [1] 2 3 4   所以return mid
3. array本身就是sorted，没有发生旋转的， 那就直接返回 nums[start] 即可     


1 2 3 4 5 6 -> 4 5 6 1 2 3        mid = 6, start ~ mid sorted 那pivot一定在mid之后，我们要找的再mid之后
1 2 3 4 5 6 -> 6 1 2 3 4 5        mid = 2, mid ~ end sorted 那我们要找的在mid之前

'''


class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return -1
        
        if len(nums) == 1:
            return nums[0]
        
        start = 0
        end = len(nums) - 1
        
        while(start <= end):
            mid = (start + end) // 2
            midNum = nums[mid]
            
            #case 1 数组本来就是sorted 并没有发生旋转，那就直接return 第一个
            if (midNum >= nums[start] and midNum <= nums[end]):
                return nums[start]
            
            #case 2 check if mid is already a pivot 
            if (mid - 1 >= 0 and mid + 1 < len(nums) and midNum < nums[mid- 1] and midNum < nums[mid + 1]):
                return midNum
            
            #case 3 midnumber后一位小于了midNum，那后一位一定是最小的，因为如果数组本身是sorted， 它后面的一定是大的，但现在小了，就说明后面一位是pivot
            if (mid + 1 < len(nums) and nums[mid + 1] < midNum):
                return nums[mid + 1]
            
            #case 2, start ~ mid sorted
            if (midNum >= nums[start]):
                #check                
                start = mid + 1
            
            else:
                end = mid - 1
        
        return -1
        
        